packages/terminal-external/src/electron-browser/terminal-external-contribution.ts

```diff
@@ -0,0 +1,128 @@
+
+import { inject, injectable } from 'inversify';
+import {
+    Command,
+    CommandContribution,
+    CommandRegistry
+} from '@theia/core/lib/common';
+import { EnvVariablesServer } from '@theia/core/lib/common/env-variables';
+import { QuickPickService } from '@theia/core/lib/common/quick-pick-service';
+import {
+    KeybindingContribution,
+    KeybindingRegistry,
+    LabelProvider
+} from '@theia/core/lib/browser';
+import { EditorManager } from '@theia/editor/lib/browser/editor-manager';
+import { WorkspaceService } from '@theia/workspace/lib/browser';
+import { TerminalExternalService } from '../common/terminal-external';
+import { TerminalExternalPreferenceService } from './terminal-external-preference';
+
+export namespace TerminalExternalCommands {
+    export const OPEN_NATIVE_CONSOLE: Command = {
+        id: 'workbench.action.terminal.openNativeConsole',
+        label: 'Open New External Terminal'
+    };
+}
+
+@injectable()
+export class TerminalExternalFrontendContribution implements CommandContribution, KeybindingContribution {
+
+    @inject(EditorManager)
+    private readonly editorManager: EditorManager;
+
+    @inject(EnvVariablesServer)
+    private readonly envVariablesServer: EnvVariablesServer;
+
+    @inject(LabelProvider)
+    private readonly labelProvider: LabelProvider;
+
+    @inject(QuickPickService)
+    private readonly quickPickService: QuickPickService;
+
+    @inject(TerminalExternalService)
+    private readonly terminalExternalService: TerminalExternalService;
+
+    @inject(TerminalExternalPreferenceService)
+    private readonly terminalExternalPreferences: TerminalExternalPreferenceService;
+
+    @inject(WorkspaceService)
+    private readonly workspaceService: WorkspaceService;
+
+    registerCommands(commands: CommandRegistry): void {
+        commands.registerCommand(TerminalExternalCommands.OPEN_NATIVE_CONSOLE, {
+            execute: () => this.openTerminalExternal()
+        });
+    }
+
+    registerKeybindings(keybindings: KeybindingRegistry): void {
+        keybindings.registerKeybinding({
+            command: TerminalExternalCommands.OPEN_NATIVE_CONSOLE.id,
+            keybinding: 'shift+ctrlcmd+c'
+        });
+    }
+
+    /**
+     * Open native console on host machine.
+     */
+    protected async openTerminalExternal(): Promise<void> {
+        const configuration = this.terminalExternalPreferences.getTerminalExternalConfiguration();
+
+        // Multi-root workspace opened.
+        if (this.workspaceService.isMultiRootWorkspaceOpened) {
+            // Display a quick pick to let users choose which workspace to spawn the terminal.
+            const chosenWorkspaceRoot = await this.selectTerminalExternalCwd();
+            if (chosenWorkspaceRoot) {
+                await this.terminalExternalService.openTerminal(configuration, chosenWorkspaceRoot);
+            }
+            return;
+        }
+
+        // Only one workspace opened.
```

Comment:
Same as `multi-root` comment. It might be better to document the method instead to state that it supports multi-root, workspace, and fallsback to editor or user home.